Client/Frontend/Browser/TabManager.swift

```diff
@@ -509,13 +501,120 @@ class TabManager: NSObject, FeatureFlagsProtocol {
+                                 didClearTabs: @escaping (_ tabsToRemove: [Tab],
+                                                          _ isPrivate: Bool,
+                                                          _ previousTabUUID: String) -> Void) {
+        let previousSelectedTabUUID = selectedTab?.tabUUID ?? ""
+        // moved closing of multiple tabs to background thread
+        DispatchQueue.global(qos: .background).async { [unowned self] in
+
+            let tabsToRemove = isPrivate ? self.privateTabs : self.normalTabs
+
+            if isPrivate && self.privateTabs.count < 1 {
+                //Bugzilla 1646756: close last private tab clears the WKWebViewConfiguration (#6827)
+                DispatchQueue.main.async { [unowned self] in
+                    self.privateConfiguration = TabManager.makeWebViewConfig(isPrivate: true,
+                                                                             prefs: self.profile.prefs)
+                }
+            }
+
+            // clear Tabs from the list that we need to remove
+            self.tabs = self.tabs.filter { !tabsToRemove.contains($0) }
+
+            // update tab manager count
+            DispatchQueue.main.async { [unowned self] in
+                self.delegates.forEach { $0.get()?.tabManagerUpdateCount?() }
+            }
+
+            DispatchQueue.main.async { [unowned self] in
+                // after closing all normal tabs we should add a normal tab
+                if self.normalTabs.isEmpty {
+                    self.selectTab(self.addTab())
+                    storeChanges()
+                }
+            }
+
+            DispatchQueue.main.async {
+                didClearTabs(tabsToRemove, isPrivate, previousSelectedTabUUID)
+            }
+        }
+    }
+
+    func makeToastFromRecentlyClosedUrls(_ recentlyClosedTabs: [Tab], isPrivate: Bool,
+                                         previousTabUUID: String) {
+        var toast: ButtonToast?
+        let numberOfTabs = recentlyClosedTabs.count
+        if numberOfTabs > 0 {
+            var didPressButton = false
+            toast = ButtonToast(labelText:
+                    String.localizedStringWithFormat(.TabsDeleteAllUndoTitle, numberOfTabs),
+                    buttonText: .TabsDeleteAllUndoAction) { buttonPressed in
+                if buttonPressed {
+                    self.reAddTabs(tabsToAdd: recentlyClosedTabs, previousTabUUID: previousTabUUID)
+                    NotificationCenter.default.post(name: .DidTapUndoCloseAllTabToast, object: nil)
+                }
+                didPressButton = true
+            } autoDismissCompletion: {
+                guard !didPressButton else { return }
+                DispatchQueue.global(qos: .background).async { [unowned self] in
+                    let previousTab = tabs.filter { $0.tabUUID == previousTabUUID }.first
+                    self.cleanupClosedTabs(recentlyClosedTabs, previous: previousTab,
```

Comment:
nit: we should either split all the arguments or none. Mixing leads to weird looking text. haha

Reply:
oops, good catch